src/centralserver/admin-service/src/main/java/org/niis/xroad/centralserver/restapi/service/SystemParameterService.java

```diff
@@ -0,0 +1,133 @@
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Service;
+
+import javax.xml.crypto.dsig.DigestMethod;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+@Slf4j
+@Service
+@PreAuthorize("isAuthenticated()")
+@RequiredArgsConstructor
+
+/**
+ *    Class for handling SystemParameter taking HA-setup into account
+ */
+public class SystemParameterService {
+
+    public static final String INSTANCE_IDENTIFIER = "instanceIdentifier";
+    public static final String CENTRAL_SERVER_ADDRESS = "centralServerAddress";
+    public static final String AUTH_CERT_REG_URL = "authCertRegUrl";
+    public static final String DEFAULT_AUTH_CERT_REG_URL = "https://%{centralServerAddress}:4001/managementservice/";
+    public static final String CONF_HASH_ALGO_URI = "confHashAlgoUri";
+    public static final String DEFAULT_CONF_HASH_ALGO_URI = DigestMethod.SHA512;
+    public static final String CONF_SIGN_CERT_HASH_ALGO_URI = "confSignCertHashAlgoUri";
+    public static final String DEFAULT_CONF_SIGN_CERT_HASH_ALGO_URI = DigestMethod.SHA512;
+    public static final String SECURITY_SERVER_OWNERS_GROUP = "securityServerOwnersGroup";
+    public static final String DEFAULT_SECURITY_SERVER_OWNERS_GROUP = "security-server-owners";
+    public static final String DEFAULT_SECURITY_SERVER_OWNERS_GROUP_DESC = "Security server owners";
+    public static final String CONF_SIGN_DIGEST_ALGO_ID = "confSignDigestAlgoId";
+    public static final String DEFAULT_CONF_SIGN_DIGEST_ALGO_ID = CryptoUtils.SHA512_ID;
+    public static final String OCSP_FRESHNESS_SECONDS = "ocspFreshnessSeconds";
+    public static final Integer DEFAULT_OCSP_FRESHNESS_SECONDS = 3600;
+    public static final String TIME_STAMPING_INTERVAL_SECONDS = "timeStampingIntervalSeconds";
+    public static final Integer DEFAULT_TIME_STAMPING_INTERVAL_SECONDS = 60;
+    public static final String CONF_EXPIRE_INTERVAL_SECONDS = "confExpireIntervalSeconds";
+    public static final Integer DEFAULT_CONF_EXPIRE_INTERVAL_SECONDS = 600;
+    private static final String[] NODE_LOCAL_PARAMETERS = {CENTRAL_SERVER_ADDRESS};
+
+    private final SystemParameterRepository systemParameterRepository;
+    private final HAConfigStatus currentHaConfigStatus;
+
+    public String getParameterValue(String key, String defaultValue) {
+        Optional<SystemParameter> valueInDb;
+        log.debug("getParameterValue() - getting value for key:{} with default value:{}", key, defaultValue);
+        if (currentHaConfigStatus.isHaConfigured()
+                && isNodeLocalParameter(key)
+        ) {
+            valueInDb = systemParameterRepository.findByKeyAndHaNodeName(
+                    key,
+                    currentHaConfigStatus.getCurrentHaNodeName()
+            );
+        } else {
+            List<SystemParameter> systemParameters = systemParameterRepository.findByKey(key);
```

Comment:
If logically at most one system parameter with the given name should exist, the repository could already return Optional<SystemParameter> and not a list.

Reply:
No database rules nor Entities are limiting duplicate keys. This table has surrogate primary key --> not limit from that. 

This SystemParameterService mostly ensures that no duplicates appears, I see only possibilities if there are multiple users of the database and some simultaneous transactions happening to create an entry to the same key. ??? Very unlikely, but possible.

Reply:
The possibility for an anomaly (duplicate values) feels like a bug and we should fix it. But, lets create a new issue since the same problem is present in the old version, and a database migration needs to take into account that there might be duplicates.